instrumentation/kafka-clients/src/main/java/brave/kafka/clients/TracingProducer.java

```diff
@@ -60,17 +63,26 @@
     return this.send(record, null);
   }
 
-  /** We wrap the send method to add tracing. */
-  @Override
-  public Future<RecordMetadata> send(ProducerRecord<K, V> record, @Nullable Callback callback) {
-    Span span = tracing.tracer().nextSpan();
+  /**
+   * We wrap the send method to add tracing.
+   */
+  @Override public Future<RecordMetadata> send(ProducerRecord<K, V> record, @Nullable Callback callback) {
+    TraceContextOrSamplingFlags traceContextOrSamplingFlags = extractor.extract(record.headers());
```

Comment:
hmm why are we doing this for producer? This adds overhead...

Reply:
Agree, but the issue is when `input-topic -> process -> output-topic`, when producing records to `output-topic` we need to link it with the trace from consumer. This is the easiest approach I found to do it, but happy to re-evaluate the approach.

Reply:
in the previous code we require the processor to use a utility (nextSpan I think) to compose this. There was a fairly lengthy chat on gitter about this with @ImFlog so want to see where he got so that we can concentrate efforts.

Reply:
In Kafka Clients we have this:
```java
  public Span nextSpan(ConsumerRecord<?, ?> record) {
    TraceContextOrSamplingFlags extracted = extractAndClearHeaders(record);
    Span result = tracing.tracer().nextSpan(extracted);
    if (extracted.context() == null && !result.isNoop()) {
      addTags(record, result);
    }
    return result;
  }
```
The idea is to create a span when we really start to treat records in Kafka, this had to be manually done.

If we keep the same idea, we need to call this somewhere before the first transformation on the stream, so probably a transformer to call after `builder.stream(XXX)`.

@jeqo you know well the KStream API, is there anything else than a Transformer that we could use to reduce the work for users ?

Reply:
In a second thought, extending the Producer could not be an option, so instead of validating context from headers on the producer side always, we can only do it when there is no current span available.